src/core/qgslayerdefinition.cpp

```diff
@@ -66,16 +66,11 @@ bool QgsLayerDefinition::loadLayerDefinition( QDomDocument doc, QgsLayerTreeGrou
       clonedSorted << node.cloneNode();
     }
     QDomNode layersNode = doc.elementsByTagName( "maplayers" ).at( 0 );
-    // remove old children
+    // replace old children with new ones
     QDomNodeList childNodes = layersNode.childNodes();
     for ( int i = 0; i < childNodes.size(); i++ )
     {
-      layersNode.removeChild( childNodes.at( i ) );
```

Comment:
`QDomNodeList` is __live__, so [whenever you change the underlying document, the contents of the list will get updated](http://doc.qt.io/qt-4.8/qdomnodelist.html#details). The effect here was that only the first half of the children were actually removed, because `layersNode.removeChild()` reduced `childNodes.size()`.